web/packages/shared/components/TerminalSearch/TerminalSearch.tsx

```diff
@@ -0,0 +1,196 @@
+ */
+
+import { useEffect, useState, useRef, useCallback } from 'react';
+import styled, { useTheme } from 'styled-components';
+import { Flex, ButtonIcon, Box, P2, Input } from 'design';
+import { ChevronDown, ChevronUp, Cross } from 'design/Icon';
+import { SearchAddon } from '@xterm/addon-search';
+
+export interface TerminalSeacher {
+  getSearchAddon(): SearchAddon;
+  term: {
+    focus(): void;
+    attachCustomKeyEventHandler: (
+      customEvent: (e: KeyboardEvent) => boolean
+    ) => void;
+  };
+}
+
+export const TerminalSearch = ({
+  xTerm,
+  show,
+  onClose,
+  onOpen,
+  isSearchKeyboardEvent,
+}: {
+  xTerm: TerminalSeacher;
+  show: boolean;
+  onClose(): void;
+  onOpen(): void;
+  isSearchKeyboardEvent: (e: KeyboardEvent) => boolean;
+}) => {
+  const theme = useTheme();
+  const searchInputRef = useRef<HTMLInputElement>();
+  const [searchValue, setSearchValue] = useState('');
+  const [searchResults, setSearchResults] = useState<{
+    resultIndex: number;
+    resultCount: number;
+  }>({ resultIndex: 0, resultCount: 0 });
+
+  useEffect(() => {
+    xTerm.getSearchAddon().onDidChangeResults(setSearchResults);
+  }, [xTerm]);
+
+  const search = (value: string, direction: 'next' | 'previous') => {
+    const match = theme.colors.terminal.brightYellow;
+    const activeMatch = theme.colors.terminal.yellow;
+    setSearchValue(value);
+    const opts = {
+      regex: true,
+      caseSensitive: false,
+      decorations: {
+        matchOverviewRuler: match,
+        activeMatchColorOverviewRuler: activeMatch,
+        matchBackground: match,
+        activeMatchBackground: activeMatch,
+      },
+    };
+
+    if (direction === 'next') {
+      xTerm.getSearchAddon().findNext(value, opts);
+    } else {
+      xTerm.getSearchAddon().findPrevious(value, opts);
+    }
+  };
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    search(e.target.value, 'next');
+  };
+
+  const onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
+      search(e.currentTarget.value, 'next');
+    }
+    // this is if you want to close the search bar with the search input focused
+    if (e.key === 'Escape') {
+      onEscape();
+      e.preventDefault();
+      return false;
+    }
+  };
+
+  const searchNext = () => {
+    search(searchInputRef.current.value, 'next');
+  };
+
+  const searchPrevious = () => {
+    search(searchInputRef.current.value, 'previous');
+  };
+
+  const onEscape = useCallback(() => {
+    onClose();
+    xTerm.getSearchAddon().clearDecorations();
+    xTerm.term.focus();
+  }, [onClose, xTerm]);
+
+  useEffect(() => {
+    xTerm.term.attachCustomKeyEventHandler(e => {
+      if (isSearchKeyboardEvent(e)) {
+        onOpen();
+        searchInputRef.current?.focus();
+        e.preventDefault();
+        return false;
+      }
+      // this escape is handled if pressing escape with the term focused
+      if (e.key === 'Escape') {
```

Comment:
Ah i see. Ok, I've removed it from `register` and now only check in `onKeyDown` for the search input. also, I've remove dthe `show` check on that because `onKeyDown` can only be called when the input is showing anyway. Let me know if it's good now 👍